// @vue-macros/define-prop/dist/chunk-MPQAS7FU.js
import {
  code_default,
  helperId,
  transformDefineProp
} from "./chunk-CMZOREPZ.js";

// src/index.ts
import process from "node:process";
import { resolveDtsHMR } from "@vue-macros/api";
import {
  createFilter,
  detectVueVersion,
  FilterFileType,
  getFilterPattern,
  normalizePath
} from "@vue-macros/common";
import {
  createUnplugin
} from "unplugin";
function resolveOptions(options, framework) {
  const version = options.version || detectVueVersion();
  const include = getFilterPattern(
    [FilterFileType.VUE_SFC_WITH_SETUP, FilterFileType.SETUP_SFC],
    framework
  );
  return {
    include,
    isProduction: process.env.NODE_ENV === "production",
    edition: "kevinEdition",
    ...options,
    version
  };
}
var name = "unplugin-vue-define-prop";
var plugin = createUnplugin(
  (userOptions = {}, { framework }) => {
    const options = resolveOptions(userOptions, framework);
    const filter = createFilter(options);
    return {
      name,
      enforce: "pre",
      resolveId(id) {
        if (id === normalizePath(helperId)) return id;
      },
      loadInclude(id) {
        return normalizePath(id) === helperId;
      },
      load(id) {
        if (normalizePath(id) === helperId) return code_default;
      },
      transformInclude: filter,
      transform(code, id) {
        return transformDefineProp(
          code,
          id,
          options.edition,
          options.isProduction
        );
      },
      vite: {
        configResolved(config) {
          options.isProduction ??= config.isProduction;
        },
        handleHotUpdate: resolveDtsHMR
      }
    };
  }
);
var index_default = plugin;

export {
  index_default
};
