// @vue-macros/volar/dist/index.js
import {
  script_lang_default
} from "./chunk-SVTEX2AH.js";
import {
  script_sfc_default
} from "./chunk-TIVGVSKW.js";
import {
  setup_jsdoc_default
} from "./chunk-CE7IUDUJ.js";
import {
  setup_sfc_default
} from "./chunk-AQZSXO4L.js";
import {
  short_bind_default
} from "./chunk-QD2UMKIG.js";
import {
  short_vmodel_default
} from "./chunk-7C227WUK.js";
import {
  define_props_default
} from "./chunk-VEOREXRB.js";
import {
  define_slots_default
} from "./chunk-62I2GVBN.js";
import {
  export_expose_default
} from "./chunk-AVZLSJS3.js";
import {
  export_props_default
} from "./chunk-EESTJNJN.js";
import {
  export_render_default
} from "./chunk-3DUK5RPZ.js";
import {
  jsx_directive_default
} from "./chunk-4VOODELA.js";
import {
  jsx_ref_default
} from "./chunk-UROWODEW.js";
import {
  boolean_prop_default
} from "./chunk-LY4PUI2V.js";
import {
  define_emit_default
} from "./chunk-YU3SKAGH.js";
import {
  define_generic_default
} from "./chunk-5CCNBU3Q.js";
import {
  define_models_default
} from "./chunk-WQJFNYPP.js";
import {
  define_options_default
} from "./chunk-PLXCSV4I.js";
import {
  define_prop_default
} from "./chunk-FD5JG5RT.js";
import {
  getVolarOptions
} from "./chunk-DCLIBDVT.js";
import {
  define_props_refs_default
} from "./chunk-NWLMJ33B.js";

// src/index.ts
var plugins = {
  defineOptions: define_options_default,
  defineModels: define_models_default,
  defineProps: define_props_default,
  definePropsRefs: define_props_refs_default,
  shortBind: short_bind_default,
  shortVmodel: short_vmodel_default,
  defineSlots: define_slots_default,
  jsxDirective: jsx_directive_default,
  booleanProp: boolean_prop_default,
  exportRender: export_render_default,
  exportProps: export_props_default,
  exportExpose: export_expose_default,
  defineProp: define_prop_default,
  defineEmit: define_emit_default,
  defineGeneric: define_generic_default,
  setupJsdoc: setup_jsdoc_default,
  setupSFC: setup_sfc_default,
  scriptSFC: script_sfc_default,
  scriptLang: script_lang_default,
  jsxRef: jsx_ref_default
};
var plugin = (ctx) => Object.entries(plugins).flatMap(([name, plugin2]) => {
  const options = getVolarOptions(ctx, name);
  if (!options) return [];
  (ctx.vueCompilerOptions.vueMacros ??= {})[name] ??= options;
  return plugin2(ctx, options);
});
var index_default = plugin;
export {
  index_default as default,
  plugin as "module.exports"
};
